text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+---
+rfc pr: [#xxx](https://github.com/aws/aws-cdk-rfcs/pull/xxx) <-- fill this after you've already created the PR
+tracking issue: https://github.com/aws/aws-cdk-rfcs/issues/322
+---
+
+# CDK Pipelines Updated API
+
+> We are reworking the internals of CDK Pipelines to separate out the concerns of CDK Deployments and AWS CodePipeline
+> better, and provide a more streamlined user-facing API,
+> both while staying backwards compatible.
+
+## Working Backwards
+
+The final API for CDK Pipelines is finally here. Compared to the previous API:
+
+- CDK applications could not be deployed in waves. They can now, as we automatically establish runorders between
+  actions.
+- You could not build pipelines that used multiple sources to feed into the synth, or that had a custom build step. You can now.
+- Large CDK applications (> 25 stacks) could not previously be deployed because they would exceed the 50 actions per
+  stage limit that CodePipeline has. With the new implementation, the actions will automatically spread over multiple
+  stages, even if you deploy multiple applications in parallel.
+- For simple use cases, you no longer need to manage CodePipeline Artifacts: artifact management is implicit
+  as objects that produce artifacts can be used to reference them.
+
+Best of all: your old code--even though it uses no longer recommended idioms--still continues to work, so you
+can upgrade to the new style at your own pace.
+
+The following:
+
+```ts
+const sourceArtifact = new codepipeline.Artifact();
+const cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
+const integTestArtifact = new codepipeline.Artifact('IntegTests');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  cloudAssemblyArtifact,
+
+  // Where the source can be found
+  sourceAction: new codepipeline_actions.GitHubSourceAction({
+    actionName: 'GitHub',
+    output: sourceArtifact,
+    oauthToken: SecretValue.secretsManager('github-token'),
+    owner: 'OWNER'
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = pipeline.addApplicationStage(new MyStage(this, 'PreProd', {
+  env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
+}));
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    actionName: 'UseSource',
+    commands: [
+      // Comes from source
+      'cat README.md',
+    ],
+    additionalArtifacts: [sourceArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+const source = cdkp.Source.gitHub('OWNER/REPO');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  synthAction: cdkp.Build.standardNpmSynth({
```

Comment:
I really don't like the name `synthAction`. Maybe just `build`? (the class is also called `Build`).

Also, do we really need this `standardNpmSynth` thing? Isn't it just `[ 'npm install`, `npm run build` ]? If that's all there is to it, I think it's a redundant abstraction.

> I recently ran into the term ["mechanical sympathy"](https://wa.aws.amazon.com/wellarchitected/2020-07-02T19-33-23/wat.concept.mechanical-sympathy.en.html), which I think applies here. The way I'm thinking about it that sometimes over-abstracting hides details from the user that makes it hard for them to understand how the system behaves.